packages/vite/src/node/server/ws.ts

```diff
@@ -28,9 +28,10 @@ export function createWebSocketServer(
   let httpsServer: Server | undefined = undefined
 
   const hmr = isObject(config.server.hmr) && config.server.hmr
-  const wsServer =
-    (hmr && hmr.server) ||
-    ((!(hmr && hmr.port) || hmr.port !== config.server.port) && server)
+  const hmrServer = hmr && hmr.server
+  const hmrPort = hmr && hmr.port
```

Comment:
Why not set default `hmr.port` value here?
```suggestion
  const hmrPort = hmr && hmr.port || 24678
```

Reply:
@benmccann @aleclarson after this change, aren't we always creating a new server for HMR? I think the idea was to reuse the server except when the user explicitly sets an `hmr.server` or `hmr.port`

Reply:
yes, I think you're right that it changed the logic. I've reverted the last change

Reply:
I'm thoroughly confused all around at this point. Ignoring for a minute whether they're the same, I wonder what the intended functioning is. The `if (wsServer)` block that reuses an existing server - is that supposed to be called when the ports are the same or different? Both versions seem to check that the ports are different before reusing the existing server. But I'm not sure that makes sense? The configured `hmr.port` never ends up getting used in that case... Should that check be reversed?

@Conduitry I'd love an extra set of eyes and any help testing from you as well

Reply:
If I understand correctly,
- the default port `24678` is [only used in middleware mode](https://github.com/vitejs/vite/blob/0535c7077b66e735217b790749a1a6fb4d58a44e/packages/vite/src/node/plugins/clientInjections.ts#L30-L31). So assigning it at the top of this function could be confusing. Let's not do this. Adding a comment to avoid future confusion would be ideal.
- reuse the server only if the ports are the same - we need a `new WebSocketServer` call when a custom port is given;
- Many `||` and `&&` can be replaced by `??` and `?.` to improve readability.

Reply:
we can't replace `&&` with `?.` because the type of `hmr` may be `false | HmrOptions`, so we need the `&&` to coalesce the type to `HmrOptions`. I rewrote as @patak-dev suggested below, which I think is a pretty clean way to write it

Reply:
Ah, thanks! I only knew that `?.` can be used on `false` in JavaScript, but I didn't realize it could cause TypeScript issues.